lahja/endpoint.py

```diff
@@ -247,38 +296,25 @@ def _throw_if_already_connected(self, *endpoints: ConnectionConfig) -> None:
             (item, config) = await self._internal_queue.get()
             self._process_item(item, config)
 
-    def _create_external_api(self, ipc_path: pathlib.Path) -> None:
-
-        receiver = EndpointConnector(self)
-
-        class ConnectorManager(BaseManager):
-            pass
-
-        ConnectorManager.register('get_connector', callable=lambda: receiver)  # type: ignore
-
-        manager = ConnectorManager(address=str(ipc_path))  # type: ignore
-        server = manager.get_server()   # type: ignore
-        threading.Thread(target=server.serve_forever, daemon=True).start()
-
     def connect_to_endpoints_blocking(self, *endpoints: ConnectionConfig, timeout: int=30) -> None:
         """
         Connect to the given endpoints and block until the connection to every endpoint is
         established. Raises a ``TimeoutError`` if connections do not become available within
         ``timeout`` seconds (default 30 seconds).
         """
-        self._throw_if_already_connected(*endpoints)
-        for endpoint in endpoints:
-            wait_for_path_blocking(endpoint.path, timeout)
-            self._connect_if_not_already_connected(endpoint)
+        raise NotImplementedError()
 
     async def connect_to_endpoints(self, *endpoints: ConnectionConfig) -> None:
         """
         Connect to the given endpoints and await until all connections are established.
         """
         self._throw_if_already_connected(*endpoints)
+        loop = None
+        if self._loop:
+            loop = self.event_loop
```

Comment:
Maybe this is still in the "make it run" phase but I just wanted to point out that this looks suspicious to me. *Locally* falling back to `loop=None` if `self._loop` isn't set doesn't strike me as right. Especially because any attempt to read from `self.event_loop` would throw at this time.

Reply:
Sorry, I don't understand the complaint :disappointed: The logic defaults to using the current loop (which definitely exists, because this is an async method), that lets this function be called before `start_serving`. However, if our loop exists we use that instead.

I agree this isn't very pretty! but I think it's correct?

Reply:
>The logic defaults to using the current loop (which definitely exists, because this is an async method), that lets this function be called before `start_serving`.

But isn't that potentially problematic? What if it uses the default loop but then `start_serving` is called with a different one (e.g. a special one in another thread).

Reply:
Okay, good point, this is a little dangerous. It'd be nice to have some way of enforcing that Endpoint is always called from the same loop.

Reply:
Coming back to this, I sat down to try to change it but I'm not sure what to do here. I think it's valuable to be able to connect to remote endpoints without first serving, so either of them could be called first. There doesn't seem to be any good reason to move an Endpoint between event loops though.

 It seems like the safest thing is to, whenever a method is called:
- If `self._loop` is not already set, set it
And either one of:
- If `self._loop` is already set, use that loop
- If `self._loop` is already set, raise an exception if it is not equal to `asyncio.get_event_loop()`

What do you think?

Reply:
>If `self._loop` is already set, raise an exception if it is not equal to `asyncio.get_event_loop()`

That would make the most sense to me!

Reply:
Okay, I wrote a `check_event_loop` decorator which guards functions which use the loop and disallows dangerous calls.